core/bin/dust_api.rs

```diff
@@ -1156,6 +1157,126 @@ async fn data_sources_search(
     }
 }
 
+/// Update tags of a document in a data source.
+
+#[derive(serde::Deserialize)]
+struct DataSourcesDocumentsUpdateTagsPayload {
+    add_tags: Option<Vec<String>>,
+    remove_tags: Option<Vec<String>>,
+}
+
+async fn data_sources_documents_update_tags(
+    extract::Path((project_id, data_source_id, document_id)): extract::Path<(i64, String, String)>,
+    extract::Json(payload): extract::Json<DataSourcesDocumentsUpdateTagsPayload>,
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+    let add_tags = match payload.add_tags {
+        Some(tags) => tags,
+        None => vec![],
+    };
+    let remove_tags = match payload.remove_tags {
+        Some(tags) => tags,
+        None => vec![],
+    };
+    let add_tags_set: HashSet<String> = add_tags.iter().cloned().collect();
+    if add_tags.len() != add_tags_set.len() {
+        return (
+            StatusCode::BAD_REQUEST,
+            Json(APIResponse {
+                error: Some(APIError {
+                    code: String::from("bad_request"),
+                    message: String::from("The `add_tags` parameter contains duplicates."),
+                }),
+                response: None,
+            }),
+        );
+    }
+    let remove_tags_set: HashSet<String> = remove_tags.iter().cloned().collect();
+    if remove_tags.len() != remove_tags_set.len() {
+        return (
+            StatusCode::BAD_REQUEST,
+            Json(APIResponse {
+                error: Some(APIError {
+                    code: String::from("bad_request"),
+                    message: String::from("The `remove_tags` parameter contains duplicates."),
+                }),
+                response: None,
+            }),
+        );
+    }
+    if add_tags_set.intersection(&remove_tags_set).count() > 0 {
+        return (
+            StatusCode::BAD_REQUEST,
+            Json(APIResponse {
+                error: Some(APIError {
+                    code: String::from("bad_request"),
+                    message: String::from(
+                        "The `add_tags` and `remove_tags` parameters contain duplicates.",
```

Comment:
not contains duplicate but rather have a non empty intersection

Reply:
Aligned, updating the wording 👍

Reply:
fixed